### ISSUE TITLE
Creating objects and undoing them still increments the next object ID

### ISSUE THREAD

**Issue by Macksaur** [mapeditor/tiled#3267]
### Version
Tiled 1.8

### Steps to reproduce
- Insert an arbitrary rectangle
- Note its ID
- Press `undo`
- Insert an arbitrary rectangle
- Note its ID has incremented by one

---

I know this is a very low priority issue and IDs are basically "infinite" but I hate wasting them when I'm playing around and doodling before going into "the actual work" of building the level. But anyway... if Tiled is undoing the creation of an object then it knows that nothing has yet happened that will reference that ID -> ID should be safe to reuse, yes?

**Comment by bjorn** [comment]
> If Tiled is undoing the creation of an object then it knows that nothing has yet happened that will reference that ID -> ID should be safe to reuse, yes?

Yeah, right now that is true. It will not be true, when we allow object references to refer to a different map. In that case, you might prefer a broken object reference over one that looks valid but refers to a completely different object.

Of course, another fix would be to use random object IDs instead of incremental ones. Then the "next object ID" attribute will just go away. What do you think?

**Comment by Macksaur** [comment]
😬 there's always a sequence of events that ruins it...

Personally, I'm not really a fan of random IDs, however large, or however seeded/prefixed.

The current implementation is fine and it works well. Especially given how I use the IDs themselves in my game. I use the IDs to store and associate object related information such as "has this treasure chest been opened yet?". Thankfully the existing behaviour of no reuse (ever) covers this including editing and updating the level over time.

I guess will have to put up with the arbitrary gaps in IDs when I'm doodling!

**Comment by bjorn** [comment]
Did you close this out of fear I would introduce random IDs? I'm afraid being "not really being a fan" isn't very convincing to me, but I guess I could always introduce an option for this if opinions vary. :-)

**Comment by eishiya** [comment]
I'm also against random IDs. Needlessly large numbers bother me to the point where I occasionally go and edit my TMX files outside of Tiled to fix the object/layer IDs and set the nextobjectid/nextlayerid to a smaller number. Random IDs would eliminate my ability to control this, and my sanity along with it xP

Ordered IDs also make it possible to discern the approximate order objects were added in without any extra metadata, which can be useful for fixing ancient mistakes. Random IDs are also an unnecessary waste of bytes, since most of them will be many digits, unlike current IDs which slowly increase in digit count as needed.

I think the current approach with non-decreasing IDs is fine. It does the job, and while the gaps are a (minor) bother for some of us, they're less of a bother than random IDs would be. So, random IDs wouldn't actually solve any problem, and would only make the existing "problem" worse xP

**Comment by bjorn** [comment]
> So, random IDs wouldn't actually solve any problem, and would only make the existing "problem" worse xP

Well, I think they definitely solve a problem. The main one being that it becomes easier to merge changes to map files when objects have been added, which currently results in consistent conflict that are annoying to resolve (if you're lucky, worst case it  will merge fine but leave objects with the same ID in the merged result).

**Comment by Macksaur** [comment]
Yikes. I hadn't even considered merging. While I appreciate the human readable format of .tmx I personally still consider it "binary data" in my mind. Diffing software will never understand it, random IDs or not. So you can't ever expect to successfully automatically merge this data. It has to be done by hand. Every time. You either take A in whole or B in whole -> done, success. But if you really must merge two maps then it's done by hand: look at A, look at B and copy/paste what you want onto the correct trunk.

It can't work in any automatic fashion (with text based diff tools) and random IDs don't solve this.

If I then use said random IDs in engine for bookkeeping and associating data then at every merge it will also annihilate all the references and lose track of the recorded data and subsequently push the problem onto the user/designer. The random IDs don't solve anything in merges unless you also know the magic ID to keep, which is the existing problem anyway.
